src/server/utils/dates.py

```diff
@@ -67,3 +76,95 @@ def weeks_in_range(week_range: Tuple[int, int]) -> int:
         year = Year(y)
         acc += year.totalweeks()
     return acc + 1  # same week should lead to 1 week that will be queried
+
+def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) -> Optional[Sequence[Union[Tuple[int, int], int]]]:
+    """
+    Converts a mixed list of dates and date ranges to an optimized list where dates are merged into ranges where possible.
+    e.g. [20200101, 20200102, (20200101, 20200104), 20200106] -> [(20200101, 20200104), 20200106]
+    (the first two values of the original list are merged into a single range)
+    """
+    if not values or len(values) <= 1:
+        return values
+
+    # determine whether the list is of days (YYYYMMDD) or weeks (YYYYWW) based on first element
+    try:
+        if (isinstance(values[0], tuple) and guess_time_value_is_day(values[0][0]))\
+            or (isinstance(values[0], int) and guess_time_value_is_day(values[0])):
+            return days_to_ranges(values)
+        elif (isinstance(values[0], tuple) and guess_time_value_is_week(values[0][0]))\
+            or (isinstance(values[0], int) and guess_time_value_is_week(values[0])):
+            return weeks_to_ranges(values)
+        else:
+            return values
+    except:
+        return values
+
+def days_to_ranges(values: Sequence[Union[Tuple[int, int], int]]) -> Sequence[Union[Tuple[int, int], int]]:
```

Comment:
theres enough complexity and similarity in both `xs_to_ranges()` that it makes sense to refactor -- have them simply pass in handles to the encoder+decoder (`time_value_to_date`+`date_to_time_value` vs `week_value_to_week`+`week_to_time_value`) and the "increment" (`timedelta(days=1)` vs `1`) to a base method.